File: asksage-client.js
/**
 * MissionPulse AskSage Client
 * FedRAMP High / IL5 Compliant AI Provider
 * 
 * Routes CUI-marked content through AskSage for CMMC compliance
 * 
 * © 2026 Mission Meets Tech
 */

(function(global) {
  'use strict';

  // ═══════════════════════════════════════════════════════════════════
  // ASKSAGE ENDPOINT CONFIGURATION
  // ═══════════════════════════════════════════════════════════════════
  const ASKSAGE_ENDPOINTS = {
    // Commercial FedRAMP High (default)
    commercial: {
      api: 'https://api.asksage.ai',
      chat: 'https://chat.asksage.ai',
      label: 'FedRAMP High (Commercial)',
      cuiAuthorized: false
    },
    // IL5 Civilian
    il5_civilian: {
      api: 'https://api.civ.asksage.ai',
      chat: 'https://chat.civ.asksage.ai',
      label: 'IL5 Civilian (CUI Authorized)',
      cuiAuthorized: true
    },
    // IL5 DoD (requires NIPR)
    il5_dod: {
      api: 'https://api.genai.army.mil',
      chat: 'https://chat.genai.army.mil',
      label: 'IL5 DoD (NIPR Required)',
      cuiAuthorized: true
    }
  };

  // Default personas available in AskSage
  const ASKSAGE_PERSONAS = {
    1: { id: 1, name: 'Ask Sage', description: 'General purpose assistant' },
    2: { id: 2, name: 'Technical Writer', description: 'Technical documentation' },
    3: { id: 3, name: 'Proposal Writer', description: 'Government proposal writing' },
    4: { id: 4, name: 'Contract Specialist', description: 'FAR/DFARS expertise' },
    5: { id: 5, name: 'Capture Manager', description: 'BD and capture strategy' }
  };

  // ═══════════════════════════════════════════════════════════════════
  // ASKSAGE CLIENT CLASS
  // ═══════════════════════════════════════════════════════════════════
  class AskSageClient {
    constructor(config = {}) {
      this.email = config.email || null;
      this.apiKey = config.apiKey || null;
      this.accessToken = null;
      this.tokenExpiry = null;
      this.securityLevel = config.securityLevel || 'commercial';
      this.defaultModel = config.defaultModel || 'gpt-4';
      this.defaultPersona = config.defaultPersona || 1;
      this.endpoint = ASKSAGE_ENDPOINTS[this.securityLevel] || ASKSAGE_ENDPOINTS.commercial;
      
      this._listeners = {
        status: [],
        error: []
      };
    }

    // ─────────────────────────────────────────────────────────────────
    // Configuration
    // ─────────────────────────────────────────────────────────────────
    configure(config) {
      if (config.email) this.email = config.email;
      if (config.apiKey) this.apiKey = config.apiKey;
      if (config.securityLevel) {
        this.securityLevel = config.securityLevel;
        this.endpoint = ASKSAGE_ENDPOINTS[config.securityLevel] || ASKSAGE_ENDPOINTS.commercial;
      }
      if (config.defaultModel) this.defaultModel = config.defaultModel;
      if (config.defaultPersona) this.defaultPersona = config.defaultPersona;
      return this;
    }

    isConfigured() {
      return !!(this.email && this.apiKey);
    }

    isCuiAuthorized() {
      return this.endpoint?.cuiAuthorized || false;
    }

    // ─────────────────────────────────────────────────────────────────
    // Token Management (24-hour tokens)
    // ─────────────────────────────────────────────────────────────────
    async getAccessToken(forceRefresh = false) {
      // Return cached token if still valid
      if (!forceRefresh && this.accessToken && this.tokenExpiry && Date.now() < this.tokenExpiry) {
        return this.accessToken;
      }

      if (!this.email || !this.apiKey) {
        throw new Error('AskSage: Email and API key required for authentication');
      }

      try {
        const response = await fetch(`${this.endpoint.api}/user/get-token-with-api-key`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            email: this.email,
            api_key: this.apiKey
          })
        });

        if (!response.ok) {
          const error = await response.json().catch(() => ({}));
          throw new Error(error.message || `Authentication failed: ${response.status}`);
        }

        const data = await response.json();
        this.accessToken = data.access_token || data.token;
        // Token valid for 24 hours, refresh at 23 hours
        this.tokenExpiry = Date.now() + (23 * 60 * 60 * 1000);
        
        this._emit('status', { status: 'authenticated', endpoint: this.endpoint.label });
        return this.accessToken;
      } catch (error) {
        this._emit('error', { error: error.message, operation: 'getAccessToken' });
        throw error;
      }
    }

    // ─────────────────────────────────────────────────────────────────
    // Core API Methods
    // ─────────────────────────────────────────────────────────────────
    
    /**
     * Query AskSage AI
     * @param {string} message - User message/prompt
     * @param {Object} options - Query options
     * @returns {Promise<Object>} - AI response
     */
    async query(message, options = {}) {
      const token = await this.getAccessToken();
      
      const payload = {
        message: message,
        model: options.model || this.defaultModel,
        persona: options.persona || this.defaultPersona,
        temperature: options.temperature || 0.7,
        dataset: options.dataset || null,
        system_prompt: options.systemPrompt || null,
        ...options.extra
      };

      try {
        const response = await fetch(`${this.endpoint.api}/server/query`, {
          method: 'POST',
          headers: {
            'x-access-tokens': token,
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(payload)
        });

        if (!response.ok) {
          const error = await response.json().catch(() => ({}));
          throw new Error(error.message || `Query failed: ${response.status}`);
        }

        const data = await response.json();
        return {
          success: true,
          response: data.response || data.message || data,
          model: payload.model,
          persona: payload.persona,
          provider: 'asksage',
          cuiCompliant: this.isCuiAuthorized()
        };
      } catch (error) {
        this._emit('error', { error: error.message, operation: 'query' });
        throw error;
      }
    }

    /**
     * Query with file attachment
     * @param {string} message - User message
     * @param {string|string[]} files - File path(s)
     * @param {Object} options - Query options
     */
    async queryWithFile(message, files, options = {}) {
      const token = await this.getAccessToken();
      
      const payload = {
        message: message,
        file: Array.isArray(files) ? files : [files],
        model: options.model || this.defaultModel,
        persona: options.persona || this.defaultPersona,
        ...options.extra
      };

      try {
        const response = await fetch(`${this.endpoint.api}/server/query_with_file`, {
          method: 'POST',
          headers: {
            'x-access-tokens': token,
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(payload)
        });

        if (!response.ok) {
          throw new Error(`Query with file failed: ${response.status}`);
        }

        const data = await response.json();
        return {
          success: true,
          response: data.response || data,
          provider: 'asksage'
        };
      } catch (error) {
        this._emit('error', { error: error.message, operation: 'queryWithFile' });
        throw error;
      }
    }

    /**
     * Get available models
     */
    async getModels() {
      const token = await this.getAccessToken();
      
      try {
        const response = await fetch(`${this.endpoint.api}/server/get-models`, {
          method: 'POST',
          headers: {
            'x-access-tokens': token,
            'Content-Type': 'application/json'
          }
        });

        if (!response.ok) {
          throw new Error(`Get models failed: ${response.status}`);
        }

        return await response.json();
      } catch (error) {
        this._emit('error', { error: error.message, operation: 'getModels' });
        throw error;
      }
    }

    /**
     * Get available personas
     */
    async getPersonas() {
      const token = await this.getAccessToken();
      
      try {
        const response = await fetch(`${this.endpoint.api}/server/get-personas`, {
          method: 'POST',
          headers: {
            'x-access-tokens': token,
            'Content-Type': 'application/json'
          }
        });

        if (!response.ok) {
          throw new Error(`Get personas failed: ${response.status}`);
        }

        return await response.json();
      } catch (error) {
        this._emit('error', { error: error.message, operation: 'getPersonas' });
        throw error;
      }
    }

    /**
     * Get user's datasets
     */
    async getDatasets() {
      const token = await this.getAccessToken();
      
      try {
        const response = await fetch(`${this.endpoint.api}/server/get-datasets`, {
          method: 'POST',
          headers: {
            'x-access-tokens': token,
            'Content-Type': 'application/json'
          }
        });

        if (!response.ok) {
          throw new Error(`Get datasets failed: ${response.status}`);
        }

        return await response.json();
      } catch (error) {
        this._emit('error', { error: error.message, operation: 'getDatasets' });
        throw error;
      }
    }

    /**
     * Get token usage for current month
     */
    async getTokenUsage() {
      const token = await this.getAccessToken();
      
      try {
        const response = await fetch(`${this.endpoint.api}/server/count-monthly-tokens`, {
          method: 'POST',
          headers: {
            'x-access-tokens': token,
            'Content-Type': 'application/json'
          }
        });

        if (!response.ok) {
          throw new Error(`Get token usage failed: ${response.status}`);
        }

        return await response.json();
      } catch (error) {
        this._emit('error', { error: error.message, operation: 'getTokenUsage' });
        throw error;
      }
    }

    /**
     * Test connection to AskSage
     */
    async testConnection() {
      try {
        await this.getAccessToken(true);
        const models = await this.getModels();
        return {
          success: true,
          message: 'Connected to AskSage',
          endpoint: this.endpoint.label,
          cuiAuthorized: this.isCuiAuthorized(),
          modelsAvailable: Array.isArray(models) ? models.length : 'unknown'
        };
      } catch (error) {
        return {
          success: false,
          message: error.message,
          endpoint: this.endpoint.label
        };
      }
    }

    // ─────────────────────────────────────────────────────────────────
    // Event System
    // ─────────────────────────────────────────────────────────────────
    on(event, callback) {
      if (this._listeners[event]) {
        this._listeners[event].push(callback);
      }
      return this;
    }

    off(event, callback) {
      if (this._listeners[event]) {
        this._listeners[event] = this._listeners[event].filter(cb => cb !== callback);
      }
      return this;
    }

    _emit(event, data) {
      if (this._listeners[event]) {
        this._listeners[event].forEach(cb => cb(data));
      }
    }
  }

  // ═══════════════════════════════════════════════════════════════════
  // EXPORT
  // ═══════════════════════════════════════════════════════════════════
  global.AskSageClient = AskSageClient;
  global.ASKSAGE_ENDPOINTS = ASKSAGE_ENDPOINTS;
  global.ASKSAGE_PERSONAS = ASKSAGE_PERSONAS;

  console.log('[MissionPulse] AskSage Client loaded');

})(typeof window !== 'undefined' ? window : global);

File: missionpulse_asksage/asksage_client.py
"""
================================================================================
MISSIONPULSE - ASKSAGE API CLIENT
================================================================================
FedRAMP High / IL5 compliant client for AskSage API integration.
Handles authentication, token refresh, and intelligent model routing.

SECURITY NOTES:
- API keys stored in environment variables only
- Tokens expire after 24 hours (auto-refresh implemented)
- All requests use TLS 1.3
- No CUI stored locally - fire-and-forget pattern

Author: Mission Meets Tech
Version: 1.0.0
Security: NIST 800-53 Rev 5 CHECKED
Encryption: FIPS 140-2 COMPLIANT
================================================================================
"""

import os
import time
import json
import logging
import asyncio
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any, AsyncGenerator
from dataclasses import dataclass, field
from enum import Enum
import httpx

from .model_config import MODEL_REGISTRY, get_model, ModelDefinition
from .query_classifier import QueryClassifier, ClassificationResult, get_classifier

logger = logging.getLogger(__name__)


class AskSageEnvironment(str, Enum):
    """AskSage deployment environments."""
    COMMERCIAL = "commercial"      # IL2, non-CUI
    FEDRAMP_HIGH = "fedramp_high"  # IL5, CUI authorized
    DOD = "dod"                    # NIPR connection required


@dataclass
class AskSageConfig:
    """Configuration for AskSage client."""
    api_key: str
    email: str
    environment: AskSageEnvironment = AskSageEnvironment.COMMERCIAL
    
    # Timeout settings
    connect_timeout: float = 10.0
    read_timeout: float = 120.0
    
    # Retry settings
    max_retries: int = 3
    retry_delay: float = 1.0
    
    # Token management
    token_refresh_buffer_minutes: int = 60  # Refresh 1 hour before expiry
    
    @property
    def base_url(self) -> str:
        """Get base URL for configured environment."""
        urls = {
            AskSageEnvironment.COMMERCIAL: "https://api.asksage.ai",
            AskSageEnvironment.FEDRAMP_HIGH: "https://api.civ.asksage.ai",
            AskSageEnvironment.DOD: "https://api.genai.army.mil"
        }
        return urls[self.environment]
    
    @classmethod
    def from_env(cls) -> "AskSageConfig":
        """Create config from environment variables."""
        api_key = os.getenv("ASKSAGE_API_KEY")
        email = os.getenv("ASKSAGE_EMAIL")
        env_str = os.getenv("ASKSAGE_ENVIRONMENT", "commercial").lower()
        
        if not api_key:
            raise ValueError("ASKSAGE_API_KEY environment variable required")
        if not email:
            raise ValueError("ASKSAGE_EMAIL environment variable required")
        
        environment = {
            "commercial": AskSageEnvironment.COMMERCIAL,
            "fedramp_high": AskSageEnvironment.FEDRAMP_HIGH,
            "fedramp": AskSageEnvironment.FEDRAMP_HIGH,
            "high": AskSageEnvironment.FEDRAMP_HIGH,
            "dod": AskSageEnvironment.DOD,
            "nipr": AskSageEnvironment.DOD
        }.get(env_str, AskSageEnvironment.COMMERCIAL)
        
        return cls(api_key=api_key, email=email, environment=environment)


@dataclass
class TokenInfo:
    """Access token with expiration tracking."""
    token: str
    expires_at: datetime
    
    @property
    def is_expired(self) -> bool:
        return datetime.utcnow() >= self.expires_at
    
    def needs_refresh(self, buffer_minutes: int = 60) -> bool:
        """Check if token should be refreshed (with buffer)."""
        refresh_at = self.expires_at - timedelta(minutes=buffer_minutes)
        return datetime.utcnow() >= refresh_at


@dataclass
class AskSageResponse:
    """Structured response from AskSage API."""
    content: str
    model: str
    input_tokens: int
    output_tokens: int
    total_tokens: int
    estimated_cost: float
    
    # Routing metadata
    classification: Optional[ClassificationResult] = None
    
    # Response metadata
    response_time_ms: int = 0
    request_id: Optional[str] = None
    
    # Sources/citations from RAG
    sources: List[Dict[str, Any]] = field(default_factory=list)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "content": self.content,
            "model": self.model,
            "input_tokens": self.input_tokens,
            "output_tokens": self.output_tokens,
            "total_tokens": self.total_tokens,
            "estimated_cost": self.estimated_cost,
            "classification": self.classification.to_dict() if self.classification else None,
            "response_time_ms": self.response_time_ms,
            "request_id": self.request_id,
            "sources": self.sources
        }


class AskSageClient:
    """
    FedRAMP-compliant AskSage API client.
    
    Features:
    - Automatic token refresh
    - Intelligent model routing
    - Streaming support
    - CUI-safe operation (fire-and-forget)
    - Comprehensive error handling
    """
    
    def __init__(self, config: Optional[AskSageConfig] = None):
        """
        Initialize AskSage client.
        
        Args:
            config: Client configuration (from env if not provided)
        """
        self.config = config or AskSageConfig.from_env()
        self._token: Optional[TokenInfo] = None
        self._http_client: Optional[httpx.AsyncClient] = None
        
        # Initialize classifier
        force_cui = self.config.environment != AskSageEnvironment.COMMERCIAL
        self.classifier = get_classifier(force_cui_models=force_cui)
        
        logger.info(
            f"AskSage client initialized for {self.config.environment.value} "
            f"environment at {self.config.base_url}"
        )
    
    async def _get_http_client(self) -> httpx.AsyncClient:
        """Get or create HTTP client."""
        if self._http_client is None or self._http_client.is_closed:
            self._http_client = httpx.AsyncClient(
                timeout=httpx.Timeout(
                    connect=self.config.connect_timeout,
                    read=self.config.read_timeout,
                    write=30.0,
                    pool=5.0
                ),
                # FIPS 140-2: Enforce TLS 1.2+ (httpx uses system SSL)
                verify=True,
                http2=True
            )
        return self._http_client
    
    async def close(self) -> None:
        """Close HTTP client."""
        if self._http_client and not self._http_client.is_closed:
            await self._http_client.aclose()
    
    # =========================================================================
    # AUTHENTICATION
    # =========================================================================
    
    async def _ensure_token(self) -> str:
        """Ensure we have a valid access token, refreshing if needed."""
        if self._token is None or self._token.needs_refresh(
            self.config.token_refresh_buffer_minutes
        ):
            await self._refresh_token()
        return self._token.token
    
    async def _refresh_token(self) -> None:
        """Get new access token from AskSage."""
        client = await self._get_http_client()
        
        url = f"{self.config.base_url}/user/get-token-with-api-key"
        payload = {
            "email": self.config.email,
            "api_key": self.config.api_key
        }
        
        try:
            response = await client.post(
                url,
                json=payload,
                headers={"Content-Type": "application/json"}
            )
            response.raise_for_status()
            
            data = response.json()
            token = data.get("token") or data.get("access_token")
            
            if not token:
                raise ValueError("No token in response")
            
            # Tokens are valid for 24 hours
            self._token = TokenInfo(
                token=token,
                expires_at=datetime.utcnow() + timedelta(hours=24)
            )
            
            logger.info("AskSage access token refreshed successfully")
            
        except httpx.HTTPStatusError as e:
            logger.error(f"Token refresh failed: {e.response.status_code}")
            raise RuntimeError(f"AskSage authentication failed: {e.response.text}")
        except Exception as e:
            logger.error(f"Token refresh error: {e}")
            raise
    
    # =========================================================================
    # MODEL OPERATIONS
    # =========================================================================
    
    async def get_available_models(self) -> List[Dict[str, Any]]:
        """Get list of available models from AskSage."""
        token = await self._ensure_token()
        client = await self._get_http_client()
        
        url = f"{self.config.base_url}/server/get-models"
        headers = {
            "x-access-tokens": token,
            "Content-Type": "application/json"
        }
        
        response = await client.post(url, headers=headers)
        response.raise_for_status()
        
        return response.json().get("models", [])
    
    # =========================================================================
    # QUERY OPERATIONS
    # =========================================================================
    
    async def query(
        self,
        message: str,
        agent_type: str = "capture",
        model: Optional[str] = None,
        system_prompt: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None,
        conversation_history: Optional[List[Dict]] = None,
        temperature: float = 0.7,
        max_tokens: int = 4096,
        dataset: Optional[str] = None,
        persona: Optional[int] = None
    ) -> AskSageResponse:
        """
        Send query to AskSage with intelligent model routing.
        
        Args:
            message: User's query
            agent_type: Which MissionPulse agent (for routing)
            model: Override model selection (optional)
            system_prompt: System prompt for the model
            context: Additional context for routing/prompting
            conversation_history: Prior conversation messages
            temperature: Response randomness (0-1)
            max_tokens: Maximum response length
            dataset: AskSage dataset ID for RAG
            persona: AskSage persona ID
            
        Returns:
            AskSageResponse with content and metadata
        """
        start_time = time.time()
        
        # Classify query and select model if not overridden
        classification = None
        if model is None:
            classification = self.classifier.classify(
                query=message,
                agent_type=agent_type,
                context=context,
                conversation_history=conversation_history
            )
            model = classification.recommended_model
            logger.info(
                f"Query classified: score={classification.complexity_score}, "
                f"model={model}, cui={classification.cui_detected}"
            )
        
        # Get token and client
        token = await self._ensure_token()
        client = await self._get_http_client()
        
        # Build request
        url = f"{self.config.base_url}/server/query"
        headers = {
            "x-access-tokens": token,
            "Content-Type": "application/json"
        }
        
        # Build messages for conversation
        messages = []
        if conversation_history:
            for msg in conversation_history:
                messages.append({
                    "role": msg.get("role", "user"),
                    "content": msg.get("content", "")
                })
        messages.append({"role": "user", "content": message})
        
        payload: Dict[str, Any] = {
            "model": model,
            "message": message,
            "temperature": temperature,
            "max_tokens": max_tokens
        }
        
        # Add optional parameters
        if system_prompt:
            payload["system_prompt"] = system_prompt
        if dataset:
            payload["dataset"] = dataset
        if persona:
            payload["persona"] = persona
        if conversation_history:
            payload["history"] = conversation_history
        
        # Make request with retries
        last_error = None
        for attempt in range(self.config.max_retries):
            try:
                response = await client.post(url, json=payload, headers=headers)
                response.raise_for_status()
                data = response.json()
                
                # Extract response content
                content = data.get("response", data.get("message", ""))
                
                # Extract token usage
                usage = data.get("usage", {})
                input_tokens = usage.get("prompt_tokens", usage.get("input_tokens", 0))
                output_tokens = usage.get("completion_tokens", usage.get("output_tokens", 0))
                total_tokens = input_tokens + output_tokens
                
                # Calculate cost
                model_def = get_model(model)
                if model_def:
                    estimated_cost = model_def.estimated_cost(input_tokens, output_tokens)
                else:
                    # Fallback estimate
                    estimated_cost = (input_tokens * 0.003 + output_tokens * 0.015) / 1000
                
                response_time = int((time.time() - start_time) * 1000)
                
                return AskSageResponse(
                    content=content,
                    model=model,
                    input_tokens=input_tokens,
                    output_tokens=output_tokens,
                    total_tokens=total_tokens,
                    estimated_cost=estimated_cost,
                    classification=classification,
                    response_time_ms=response_time,
                    request_id=data.get("request_id"),
                    sources=data.get("sources", [])
                )
                
            except httpx.HTTPStatusError as e:
                last_error = e
                if e.response.status_code == 401:
                    # Token expired, refresh and retry
                    logger.warning("Token expired, refreshing...")
                    await self._refresh_token()
                    token = self._token.token
                    headers["x-access-tokens"] = token
                elif e.response.status_code >= 500:
                    # Server error, retry with backoff
                    logger.warning(f"Server error, retrying in {self.config.retry_delay}s...")
                    await asyncio.sleep(self.config.retry_delay * (attempt + 1))
                else:
                    raise
            except Exception as e:
                last_error = e
                logger.error(f"Query error (attempt {attempt + 1}): {e}")
                if attempt < self.config.max_retries - 1:
                    await asyncio.sleep(self.config.retry_delay)
        
        raise RuntimeError(f"Query failed after {self.config.max_retries} attempts: {last_error}")
    
    async def stream(
        self,
        message: str,
        agent_type: str = "capture",
        model: Optional[str] = None,
        system_prompt: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None,
        conversation_history: Optional[List[Dict]] = None,
        temperature: float = 0.7,
        max_tokens: int = 4096
    ) -> AsyncGenerator[str, None]:
        """
        Stream response chunks from AskSage.
        
        Note: Streaming support depends on AskSage endpoint availability.
        Falls back to non-streaming if not supported.
        
        Yields:
            Response text chunks as they arrive
        """
        # Classify and select model
        classification = None
        if model is None:
            classification = self.classifier.classify(
                query=message,
                agent_type=agent_type,
                context=context,
                conversation_history=conversation_history
            )
            model = classification.recommended_model
        
        token = await self._ensure_token()
        client = await self._get_http_client()
        
        url = f"{self.config.base_url}/server/query"
        headers = {
            "x-access-tokens": token,
            "Content-Type": "application/json",
            "Accept": "text/event-stream"
        }
        
        payload = {
            "model": model,
            "message": message,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "stream": True
        }
        
        if system_prompt:
            payload["system_prompt"] = system_prompt
        if conversation_history:
            payload["history"] = conversation_history
        
        try:
            async with client.stream("POST", url, json=payload, headers=headers) as response:
                response.raise_for_status()
                
                async for line in response.aiter_lines():
                    if line.startswith("data: "):
                        data_str = line[6:]
                        if data_str == "[DONE]":
                            break
                        try:
                            data = json.loads(data_str)
                            chunk = data.get("delta", {}).get("content", "")
                            if chunk:
                                yield chunk
                        except json.JSONDecodeError:
                            continue
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 400:
                # Streaming not supported, fall back
                logger.warning("Streaming not supported, falling back to standard query")
                response = await self.query(
                    message=message,
                    agent_type=agent_type,
                    model=model,
                    system_prompt=system_prompt,
                    context=context,
                    conversation_history=conversation_history,
                    temperature=temperature,
                    max_tokens=max_tokens
                )
                yield response.content
            else:
                raise
    
    # =========================================================================
    # CONVENIENCE METHODS
    # =========================================================================
    
    async def simple_query(self, message: str, model: str = "gpt-4o-mini") -> str:
        """Simple query without routing - returns just the text."""
        response = await self.query(message=message, model=model)
        return response.content
    
    async def cui_query(
        self,
        message: str,
        agent_type: str = "compliance",
        system_prompt: Optional[str] = None
    ) -> AskSageResponse:
        """
        Query specifically for CUI content.
        Forces CUI-authorized model regardless of classification.
        """
        # Force CUI-authorized model
        if self.config.environment == AskSageEnvironment.COMMERCIAL:
            logger.warning(
                "CUI query requested but using commercial environment. "
                "Consider switching to FedRAMP High for CUI workloads."
            )
        
        return await self.query(
            message=message,
            agent_type=agent_type,
            model="claude-sonnet-4",  # Always use CUI-authorized
            system_prompt=system_prompt,
            context={"cui_marked": True}
        )


# =============================================================================
# SINGLETON / FACTORY
# =============================================================================

_client_instance: Optional[AskSageClient] = None


def get_asksage_client(config: Optional[AskSageConfig] = None) -> AskSageClient:
    """Get or create AskSage client singleton."""
    global _client_instance
    if _client_instance is None:
        _client_instance = AskSageClient(config)
    return _client_instance


async def create_asksage_client(config: Optional[AskSageConfig] = None) -> AskSageClient:
    """Create new AskSage client (async factory)."""
    client = AskSageClient(config)
    # Warm up token
    await client._ensure_token()
    return client
